File: sdk/python/resolvetrace/identity.py
"""Per-client identity state.

Holds the in-memory ``(user_id, traits)`` pair set by ``client.identify(...)``.
Kept separate from session lifecycle so identity changes never roll over a
session.

The wire-format mapping in the current contract surfaces identity in two
places: the ``identify`` block on ``/v1/session/start`` (``{ userId, traits? }``)
and the ``actor`` decoration on every event envelope captured while an
identity is set (``{ userId, traits? }``). Both use camelCase wire keys.
``traits`` is carried as a free-form trait bag in both shapes.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class IdentitySnapshot:
    """Immutable read of the current identity, returned by :meth:`IdentityState.snapshot`."""

    user_id: str | None
    traits: dict[str, Any] | None


class IdentityState:
    """Thread-safe identity holder.

    The instance is shared between the session manager (which reads it when
    building ``/v1/session/start`` bodies) and the client (which reads it
    when constructing event envelopes).
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._user_id: str | None = None
        self._traits: dict[str, Any] | None = None

    def set(self, user_id: str | None, traits: dict[str, Any] | None = None) -> None:
        """Set or clear the active identity.

        Passing ``user_id=None`` clears identity (subsequent reads return
        ``None``). ``traits`` is stored as a shallow copy.
        """
        with self._lock:
            self._user_id = user_id
            if user_id is None:
                self._traits = None
            else:
                self._traits = dict(traits) if traits else None

    def clear(self) -> None:
        """Equivalent to ``set(None)``."""
        self.set(None)

    def snapshot(self) -> IdentitySnapshot:
        """Return a stable copy of the current identity."""
        with self._lock:
            traits_copy = dict(self._traits) if self._traits is not None else None
            return IdentitySnapshot(user_id=self._user_id, traits=traits_copy)

    @property
    def user_id(self) -> str | None:
        with self._lock:
            return self._user_id


__all__ = ["IdentityState", "IdentitySnapshot"]

File: sdk/typescript/src/identity.ts
/**
 * In-memory identity state set by `client.identify(...)`.
 *
 * Kept deliberately separate from the session manager: identity is a
 * decoration that flows through every event envelope and the next
 * `/v1/session/start` body, but it does NOT drive session lifecycle
 * (no rollover, no network call by itself).
 */

import type { ActorIdentity } from './types.js';

/** Snapshot of the current identity, or null when none is set. */
export interface IdentitySnapshot {
  userId: string;
  traits?: Record<string, unknown>;
}

/** Holds the current identity decoration in memory. */
export class IdentityState {
  private current: IdentitySnapshot | null = null;

  /** Replace the current identity. Pass `null` for `userId` to clear. */
  set(userId: string | null, traits?: Record<string, unknown>): void {
    if (userId === null) {
      this.current = null;
      return;
    }
    if (typeof userId !== 'string' || userId.length === 0) {
      throw new TypeError('identify(userId, traits?) requires a non-empty string user id or null.');
    }
    const next: IdentitySnapshot = { userId };
    if (traits !== undefined) {
      if (traits === null || typeof traits !== 'object') {
        throw new TypeError('identify(userId, traits?) requires `traits` to be an object if provided.');
      }
      next.traits = { ...traits };
    }
    this.current = next;
  }

  /** Read the current identity snapshot, or `null` if none is set. */
  get(): IdentitySnapshot | null {
    return this.current;
  }

  /** Clear the current identity. */
  clear(): void {
    this.current = null;
  }

  /**
   * Build the `actor` decoration for an event envelope, or `undefined` when
   * no identity is currently held.
   */
  toActor(): ActorIdentity | undefined {
    if (this.current === null) return undefined;
    const out: ActorIdentity = { userId: this.current.userId };
    if (this.current.traits !== undefined) {
      out.traits = { ...this.current.traits };
    }
    return out;
  }
}
